### PR TITLE
[virt_autotest] Enhancement for libvirt_virtual_network_init

### PR DESCRIPTION
**Pull request by guoxuguang**
- Enhancement for libvirt_virtual_network_init
We need to ensure the given guest system is keeping running status, the SSH connection and the ICMP PING responses is workable together at the libvirt_virtual_network_init test module as our expected. So, add the new `function validate_guest_status() `to ensure each check point as our expected for given guest system at the libvirt_virtual_network_init test module here, instead of current `script_retry()` this easy check way. 
- Related ticket: 
https://progress.opensuse.org/issues/122080
- Verification run: 
[gi-guest_developing-on-host_developing-kvm](http://10.67.129.218/tests/289)
[gi-guest_developing-on-host_developing-xen](http://10.67.129.218/tests/284)

### CODE REVIEW

File: tests/virt_autotest/libvirt_virtual_network_init.pm [hunk 1]
```diff
@@ -67,8 +67,8 @@ sub run_test {
         upload_logs "/tmp/$guest.xml";
         #Used with attach-detach(hotplugging) interface to confirm all virtual network mode
         #NOTE: Required all guests keep running status
-        #Check that all guests are still running before virtual network tests
-        script_retry("nmap $guest -PN -p ssh | grep open", delay => 30, retry => 6, timeout => 180);
+        #Ensures the given guests is started and fixes some common network issues
+        ensure_online($guest);
```
Comment (alice-suse) [inline tests/virt_autotest/libvirt_virtual_network_init.pm:71]: Thanks for the PR, Leon. This replacement should be able to continue test when guest is down/stuck unexpectedly.

 As part of the original intention for the ticket, we need to find root cause why the guest ssh port check fails. I did not find your analysis result in the progress ticket. Do I miss it somewhere? BTW, it will be helpful to add more debug info to let us know the real status of the guest at that moment, before trying to restart the guest in ensure_online. HDYT?

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm:71]: > Thanks for the PR, Leon. This replacement should be able to continue test when guest is down/stuck unexpectedly.
> 
> As part of the original intention for the ticket, we need to find root cause why the guest ssh port check fails. I did not find your analysis result in the progress ticket. Do I miss it somewhere? 
Good point,  emm.. sorry for my mistake, I have missed that.  back to [poo#122080](https://progress.opensuse.org/issues/122080), after did more investigation, figure out that there just only reproduce this libvirt_virtual_network_init test failure one time for 15-SP5 FV guest system on 15-SP5 KVM host with build61.1. meanwhile, refer to the current collect logs, confirm that TIMEOUT(i.e 180 sec) to ensure ssh port for the given 15-SP5 FV guest. meanwhile, and rerun this test get PASS results for libvirt_virtual_network_init - https://openqa.suse.de/tests/10189373#step/libvirt_virtual_network_init/1

> BTW, it will be helpful to add more debug info to let us know the real status of the guest at that moment, before trying to restart the guest in ensure_online. HDYT?
yes, this is great idea. currently,  use with ensure_online() for given guest system, which will included that check if we can ping with the given guest system, check with the ssh service connection status for the given guest system,  ensure the default route is set for the given guest system, check if we can ping SLE VM host from the guest and check also if the name resolution works for the given guest system as the default debug info. So, to be honest, I think that is enough debug info to let us know the real status of the guest at that moment during call with ensure_online() with the given guest system.

Comment (alice-suse) [reply tests/virt_autotest/libvirt_virtual_network_init.pm:71]: > > BTW, it will be helpful to add more debug info to let us know the real status of the guest at that moment, before trying to restart the guest in ensure_online. HDYT?
> 
> yes, this is great idea. currently, use with ensure_online() for given guest system, which will included that check if we can ping with the given guest system, check with the ssh service connection status for the given guest system, ensure the default route is set for the given guest system, check if we can ping SLE VM host from the guest and check also if the name resolution works for the given guest system as the default debug info. So, to be honest, I think that is enough debug info to let us know the real status of the guest at that moment during call with ensure_online() with the given guest system.

Well, I was talking something different. Maybe let me rephrase... Ensure_online() will give the guest state for ping/ssh service/route/etc. That's true. But it is the state of the VM AFTER it's restarted if not up. But before ensure_online, at this module of libvirt_virtual_network_init, the vms SHOULD BE UP AND RUNNING WELL , for which the original script_retry to check ssh port open is enough. Should not we find out why the vm is NOT in that expected state at this point and eliminate that from happening again? But if we switch to ensure_online , the issue will be hided/workarounded without any warning given. I hope the difference is  stated clearly this time?

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm:71]: Great! make sense totally as your mentioned. 
Thank you so much for your great help and suggestion. it is so helpful to me. emm, let me think about it and update code later.

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm]: @alice-suse short update, we need to ensure the given guest system is keeping **running** status, the **SSH** connection and the **ICMP PING** responses is workable together at the libvirt_virtual_network_init test module as our expected. So, add the new function `get_guest_status()` to ensure the each check point as our expected for given guest system at the libvirt_virtual_network_init test module. 
Meanwhile, update the latest validation run as below:
[gi-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/10312928#)

**Comment by guoxuguang** [comment]
@alice-suse short update, we need to ensure the given guest system is keeping **running** status, the **SSH** connection and the **ICMP PING** responses is workable together at the libvirt_virtual_network_init test module as our expected. So, add the new function `get_guest_status()` to ensure the each check point as our expected for given guest system at the libvirt_virtual_network_init test module. 
Meanwhile, update the latest validation run as below:
[gi-guest_developing-on-host_developing-kvm](https://openqa.suse.de/tests/10324498#)
[gi-guest_developing-on-host_developing-xen](http://10.67.129.218/tests/283#)

File: lib/virt_autotest/virtual_network_utils.pm [hunk 2]
```diff
@@ -362,4 +362,20 @@ sub setup_vm_simple_dns_with_ip {
     record_info("Simple DNS setup in /etc/hosts for $_ip $_vm is successful!", script_output("cat /etc/hosts"));
 }
 
+sub get_guest_status {
+    my ($guest, %args) = @_;
+    my $timeout = $args{timeout} // "180";
```
Comment (waynechen55) [inline lib/virt_autotest/virtual_network_utils.pm:367]: I do no think this parameter  my $timeout is necessary.

File: lib/virt_autotest/virtual_network_utils.pm [hunk 3]
```diff
@@ -362,4 +362,20 @@ sub setup_vm_simple_dns_with_ip {
...
+    if (script_run("virsh list --all | grep $guest | grep running") ne 0) {
+        assert_script_run "virsh list --all | grep $guest";
+        save_screenshot;
+        die "Error: $guest should keep running, please check manually!";
+    } else {
+        #Ensure the SSH connection for the given guest
+        die "Error: SSH $guest failed, please check manually!" if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 30, retry => 6, timeout => $timeout) ne 0);
+        #Ensure the ICMP PING responses for the given guest
+        die "Error: Ping $guest failed, please check manually!" if (script_retry("ping -c5 $guest", delay => 30, retry => 6, timeout => $timeout) ne 0);
+    }
```
Comment (waynechen55) [inline lib/virt_autotest/virtual_network_utils.pm:378]: From new development, I think we can start trying nc instead of nmap, because nc is installed by default, for example, https://www.moreofless.co.uk/netcat-nc-port-scanning

File: lib/virt_autotest/virtual_network_utils.pm [hunk 4]
```diff
@@ -362,4 +362,20 @@ sub setup_vm_simple_dns_with_ip {
...
+    #Ensure the given guest as running status
+    if (script_run("virsh list --all | grep $guest | grep running") ne 0) {
+        assert_script_run "virsh list --all | grep $guest";
+        save_screenshot;
+        die "Error: $guest should keep running, please check manually!";
+    } else {
+        #Ensure the SSH connection for the given guest
+        die "Error: SSH $guest failed, please check manually!" if (script_retry("nmap $guest -PN -p ssh | grep open", delay => 30, retry => 6, timeout => $timeout) ne 0);
+        #Ensure the ICMP PING responses for the given guest
+        die "Error: Ping $guest failed, please check manually!" if (script_retry("ping -c5 $guest", delay => 30, retry => 6, timeout => $timeout) ne 0);
```
Comment (waynechen55) [inline lib/virt_autotest/virtual_network_utils.pm:377]: I think you should try ping firstly followed by ssh, so more solid result is available.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 3 above)
Comment (waynechen55) [reply lib/virt_autotest/virtual_network_utils.pm:378]: namp is more powerful, but for ssh port detecting, I think nc might be enough.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 2 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm:367]: yes, good point. keep this $timeout just only to ensure future expansion for this function.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 4 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm:375]: Great, had changed to ensure ICMP PING responses firstly now.

File: lib/virt_autotest/virtual_network_utils.pm (hunk 3 above)
Comment (guoxuguang) [reply lib/virt_autotest/virtual_network_utils.pm:406]: yes, after do more investigation, figure out nc(netcat) is simple and easy tool to detect SSH port for given guest system. next, try to use nc to detect SSH instead of namp tool. meanwhile, do more local validation runs.

**Comment by guoxuguang** [comment]
Add the latest validation run:
[gi-guest_developing-on-host_developing-kvm](http://10.67.129.218/tests/289)
[gi-guest_developing-on-host_developing-xen](http://10.67.129.218/tests/284)

File: tests/virt_autotest/libvirt_virtual_network_init.pm [hunk 5]
```diff
@@ -69,8 +69,8 @@ sub run_test {
         upload_logs "/tmp/$guest.xml";
         #Used with attach-detach(hotplugging) interface to confirm all virtual network mode
         #NOTE: Required all guests keep running status
-        #Check that all guests are still running before virtual network tests
-        script_retry("nmap $guest -PN -p ssh | grep open", delay => 30, retry => 6, timeout => 180);
+        #Ensures the SSH connection and ICMP PING responses is workable for given guest system
+        get_guest_status($guest);
```
Comment (alice-suse) [inline tests/virt_autotest/libvirt_virtual_network_init.pm:73]: All looks good except that this function name seems a  little mismatch with what it does? It does not only get_*, but also assert the status with die, if I understand correctly?

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm:73]: yes, there is a little mismatch for current get_* function.  
Perhaps, the name `validate_guest_status()` would be more appropriate here.

Comment (guoxuguang) [reply tests/virt_autotest/libvirt_virtual_network_init.pm]: Quick update, had updated function name as `validate_guest_status()` now.

**Comment by guoxuguang** [comment]
Quick help, if there were new comments for this PR, please help merge it. thanks.

**Comment by alice-suse** [comment]
LGTM, if no more comments today from others, will merge it.
